packages/proto-signing/src/directsecp256k1hdwallet.ts

```diff
@@ -91,7 +91,11 @@ export class DirectSecp256k1HdWallet implements OfflineDirectSigner {
   private readonly accounts: readonly Secp256k1Derivation[];
 
   protected constructor(mnemonic: EnglishMnemonic, options: DirectSecp256k1HdWalletConstructorOptions) {
-    const { seed, hdPaths, prefix } = { ...defaultOptions, ...options };
+    const { seed, hdPaths, prefix } = {
+      seed: options.seed,
+      prefix: options.prefix ?? defaultOptions.prefix,
+      hdPaths: options.hdPaths ?? defaultOptions.hdPaths,
+    };
     this.secret = mnemonic;
     this.seed = seed;
     this.accounts = hdPaths.map((hdPath) => ({
```

Comment:
options parameter (line 93) type definition equals to:
```ts
export interface DirectSecp256k1HdWalletConstructorOptions {
  readonly bip39Password: string | undefined
  readonly hdPaths: readonly HdPath[] | undefined;
  readonly prefix: string | undefined;
  readonly seed: Uint8Array;
}
```

so it's possible to create a new instance of the `DirectSecp256k1HdWallet` class like this:
```ts
new DirectSecp256k1HdWallet(..., {
    ...
    hdPaths: undefined,
})
```

which results in a runtime TypeError on line 101
> 'Cannot read property \'map\' of undefined',

and that's because of `options` and `defaultOptions` merging with a spread operator (line 95-96):
```ts
{
    ...{
        hdPaths: [makeCosmoshubPath(0)],
    },
    ...{
        hdPaths: undefined,
    }
}
```

The typescript inferred types are misleading. One possible explicit fix is in this PR, however, there are other ways.

**functional like** (incomplete, it needs ugly casting because of filter inference):
```ts
const notUndefinedOptions = Object.fromEntries(
    Object.entries(options).filter(([_key, value]):  => value !== undefined),
)

const { seed, hdPaths, prefix } = {
  ...defaultOptions,
  ...notUndefinedOptions,
}
```

**mutate** (requires removal of `readonly`):
```ts
if (!options.hdPaths) {
  delete options.hdPaths;
}

if (!options.prefix) {
  delete options.prefix;
}

const { seed, hdPaths, prefix } = {
  ...defaultOptions,
  ...options,
}
```

Not sure which one you prefer, I like the explicit one. This is definitely an edge case but still, this is a bug, and it prevented me from a quick cosmjs upgrade to `0.25.0` in `ts-relayer`. I can make a workaround now but it's better to have it fixed anyway.

Reply:
Great explanation, thank you. Let's go for the explicit way you proposed. No need to rely on even more magic that is hard to review.

Reply:
So to wrap up what you are saying: In JavaScript we need to watch out for 3 types of null: `null`, `undefined` and not set 🤯

Reply:
Yes, exactly! This is odd... 🤯

Reply:
Ooph. Yes, this probably affects other code. :(

Reply:
@bartmacbartek Do you know why TypeScript doesn't catch this?

Reply:
I spent some time on this but no idea. I think it's worth a typescript ticket. I reproduced it here: https://www.typescriptlang.org/play?ts=4.3.0-beta#code/MYewdgzgLgBAJgUwGYEMCuAbKB5ADlAS3AgC4YBvAWACgY6YALBDDEM6AJwLAHMBtALo0AvjAC8FGvUbNWZPgHIA7iA4Y4CgQBoRNGqEiwQ+IpDJVa9JixAB+dlC69BI8ZMt1rcmGjCIk3AhwutT6xLBQCNBuFtIAdAn+6Fh4hMQ6HjAJccZpkCE0kdBxXiBxALYouAAUKOIAfDAoAJRAA